File: forge/src/main/kotlin/gay/object/caduceus/forge/ForgeCaduceus.kt
package gay.`object`.caduceus.forge

import dev.architectury.platform.forge.EventBuses
import gay.`object`.caduceus.Caduceus
import net.minecraft.data.DataProvider
import net.minecraft.data.DataProvider.Factory
import net.minecraft.data.PackOutput
import net.minecraftforge.data.event.GatherDataEvent
import net.minecraftforge.fml.common.Mod
import thedarkcolour.kotlinforforge.forge.MOD_BUS

@Mod(Caduceus.MODID)
class CaduceusForge {
    init {
        MOD_BUS.apply {
            EventBuses.registerModEventBus(Caduceus.MODID, this)
            addListener(ForgeCaduceusClient::init)
            addListener(::gatherData)
        }
        Caduceus.init()
    }

    private fun gatherData(event: GatherDataEvent) {
        event.apply {
            // TODO: add datagen providers here
        }
    }
}

fun <T : DataProvider> GatherDataEvent.addProvider(run: Boolean, factory: (PackOutput) -> T) =
    generator.addProvider(run, Factory { factory(it) })

File: forge/src/main/clojure/gay/object/caduceus/forge/ForgeCaduceus.clj
(ns gay.object.caduceus.forge.ForgeCaduceus
  (:import (dev.architectury.platform.forge EventBuses)
           (gay.object.caduceus Caduceus CaduceusClient)
           (net.minecraftforge.eventbus.api EventPriority)
           (net.minecraftforge.fml.event.lifecycle FMLClientSetupEvent)
           (net.minecraftforge.fml.javafmlmod FMLJavaModLoadingContext)))

(defn init-client [^FMLClientSetupEvent _]
  (CaduceusClient/init))

(gen-class
  :name ^{net.minecraftforge.fml.common.Mod "caduceus"}
        gay.object.caduceus.forge.ForgeCaduceus
  :prefix ForgeCaduceus-
  :init init)

(defn ForgeCaduceus-init []
  (let [bus (.getModEventBus (FMLJavaModLoadingContext/get))]
    (EventBuses/registerModEventBus Caduceus/MODID bus)
    (.addListener bus EventPriority/NORMAL false FMLClientSetupEvent init-client))
  (Caduceus/init)
  [[] nil])
